lib/pbench/server/api/resources/__init__.py

```diff
@@ -641,38 +647,88 @@ def _check_authorization(self, user: str, access: str):
+        #    data, nor any operation other than READ on PUBLIC data;
+        # 3) The "user" parameter can be omitted only for READ operations,
+        #    which will be defaulted according to the visibility of the
+        #    authenticated client.
+        # 4) An authenticated client cannot mutate data owned by a different
+        #    user, nor READ private data owned by another user.
         if self.role != API_OPERATION.READ or access == Dataset.PRIVATE_ACCESS:
+            # ROLE is UPDATE, CREATE, DELETE or ACCESS is private; we need to
+            # evaluate access rights to determine whether to allow the request
             if authorized_user is None:
+                # An unauthenticated user is never allowed to access private
+                # data nor to perform an potential mutation of data: REJECT
                 self.logger.warning(
                     "Attempt to {} user {} data without login", self.role, user
                 )
-                authorized = False
-            elif user != authorized_user.username and not authorized_user.is_admin():
+                raise UnauthorizedAccess(
+                    authorized_user, self.role, user, access, HTTPStatus.UNAUTHORIZED
+                )
+            elif self.role != API_OPERATION.READ and user is None:
+                # If no target user is specified, we won't allow any potential
+                # mutation of data: REJECT
+                self.logger.warning(
+                    "Unauthorized attempt by {} to {} data with defaulted user",
+                    authorized_user,
+                    self.role,
+                )
+                raise UnauthorizedAccess(
+                    authorized_user, self.role, user, access, HTTPStatus.FORBIDDEN
+                )
+            elif (
+                user
+                and user != authorized_user.username
+                and not authorized_user.is_admin()
+            ):
+                # If we are potentially mutating data, or reading private data,
+                # then the authenticated user must either be the owner of the
+                # data or must have ADMIN role. We won't allow the operation if
+                # these are not true: REJECT
                 self.logger.warning(
                     "Unauthorized attempt by {} to {} user {} data",
                     authorized_user,
                     self.role,
                     user,
                 )
-                authorized = False
-        if not authorized:
-            raise UnauthorizedAccess(authorized_user, self.role, user, access)
+                raise UnauthorizedAccess(
+                    authorized_user, self.role, user, access, HTTPStatus.FORBIDDEN
+                )
+            else:
+                # We have determined that there is an authenticated user with
+                # legitimate access to the specified user and access: the
```

Comment:
I would have thought you'd want this to say:

```
We have determined that there is an authenticated user with legitimate access to the specified user and access to the data: the data is public and the operation is READ, the user owns the data, or the user has ADMIN role.
```

Reply:
But not worth holding this commit for such a change.

Reply:
I was trying to avoid the apparent confusion from the reference to the dataset "user" and "access (type)" properties by simplifying the sentence. Maybe "legitimate access based on the dataset's owner and access policy" would be slightly more clear, but it's also longer and more unwieldy. I'm happy with it as is.